treefmt: refactor `withConfig` and `buildConfig` to use `evalConfig`

MattSturgeon opened the pull request:
Introduces a new modules-based implementation that is used by both `withConfig` and `buildConfig`.

Previously we used a module eval for `settings`. This PR makes it so that all args passed to `withConfig` are now module options.

`settings` is now a submodule of the wider `evalConfig` configuration. While it currently has no declared options and could just be `mkOption { inherit (settingsFormat) type; }`, keeping `settings` as a submodule-type will minimise breaking changes and allow for declaring sub-options in the future.

As well as being a better overall design (IMO), using a module eval enables adding additional options in the future. E.g. we could add `programs` options similar to those maintained by treefmt-nix. See https://github.com/numtide/treefmt-nix/issues/349

### Documentation

_EDIT: Dropped from this PR. I will address generating function & option docs in a follow up. Changes stashed [here](https://github.com/NixOS/nixpkgs/compare/31eef1e9520b0c4468362eca98296d040c9679e3...deb8ce43e5dbb6fb913f858e83357b0e401ba4eb)._

<details><summary>Details</summary>
<p>

Using the module system also enables us to generate reference docs.

I've added a section to the nixpkgs manual which includes generated reference docs for the `treefmt.evalConfig` options.

<details><summary>Screenshot</summary>
<p>

![image](https://github.com/user-attachments/assets/fbeaccea-472f-4001-901f-8a236da40f38)


</p>
</details> 

I'd like to flesh the manual section out a bit more, as currently it is effectively just a blank page to put the generated docs. Ideally, the various `treefmt.*` functions would have docs and examples, although I'm hesitant to increase the scope of this PR by writing loads of documentation.

~~I don't see a good way to include options in the nix doc-comments or include doc-comments in the manual. Maybe the doc-comments should just link to the manual?~~ EDIT: we can generate function docs using `nixdoc`.

</p>
</details> 

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - `nix-build -A treefmt.tests`
  - `nix-build -A nixfmt-tree.tests`
  - `nix-build doc`
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

cc @NixOS/nix-formatting @brianmcgee @zimbatm

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

MattSturgeon committed:
treefmt: refactor `withConfig` and `buildConfig` to use `evalConfig`

Introduces a new modules-based implementation that is used by both
`withConfig` and `buildConfig`.

Previously we used a module eval for `settings`. This PR makes it so
that all args passed to `withConfig` are now module options. `settings`
is now a submodule of the wider `evalConfig` configuration.

As well as being a better overall design (IMO), using a module eval
enables adding additional options in the future. E.g. we could add
`programs` options similar to those maintained by treefmt-nix.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/formatting-team-meeting-2025-05-13/64230/1

jfly commented on pkgs/by-name/tr/treefmt/modules/default.nix:
```diff
@@ -0,0 +1,100 @@
+
+    configFile = lib.mkOption {
+      type = lib.types.path;
+      # Ensure file is copied to the store
+      apply = file: if lib.isDerivation file then file else "${file}";
```
Won't this get copied to the store when `configFile` is referenced below?

MattSturgeon replied:
I may have missed something, but having a `"${}"` in the option's apply function feels like good practice.

This means if the option is a) defined by users to a non-store-path and b) read directly (without using the `result`), it will be coerced into a store-path.

This only really affects users who are using `treefmt.evalConfig` manually, instead of `treefmt.withConfig` or `treefmt.buildConfig`.

jfly commented on pkgs/by-name/tr/treefmt/with-config.nix:
```diff
@@ -1,32 +1,29 @@
-  ''
+
+/**
+  Wrap treefmt, configured  using structured settings.
```
```suggestion
  Wrap treefmt, configured using structured settings.
```

jfly reviewed (approved):
LGTM, but I am not an expert with the module system, nor nixpkgs's documentation.

MattSturgeon commented on pkgs/by-name/tr/treefmt/build-config.nix:
````diff
@@ -1,30 +1,32 @@
+  # Type
+
+  ```
+  Module -> Derivation
````
It can take either a single module or a list of modules. However I'm currently leaning towards list-of-modules being an undocumented/internal behaviour in case we decide to make breaking changes.

MattSturgeon commented on pkgs/by-name/tr/treefmt/modules/default.nix:
```diff
@@ -0,0 +1,100 @@
+    result =
+      pkgs.runCommand config.name
+        {
+          nativeBuildInputs = [ pkgs.makeBinaryWrapper ];
```
Documented here: https://nixos.org/manual/nixpkgs/unstable/#fun-makeWrapper

> Using the `makeBinaryWrapper` implementation is usually preferred, as it creates a tiny _compiled_ wrapper executable, that can be used as a shebang interpreter. This is needed mostly on Darwin, where shebangs cannot point to scripts, [due to a limitation with the `execve`-syscall](https://stackoverflow.com/questions/67100831/macos-shebang-with-absolute-path-not-working).
>
> Compiled wrappers generated by `makeBinaryWrapper` can be inspected with `less <path-to-wrapper>` - by scrolling past the binary data you should be able to see the shell command that generated the executable and there see the environment variables that were injected into the wrapper.

TL;DR using a wrapper script vs binary wrapper is a trade off between _readability_, _performance_, and better _darwin support_.

MattSturgeon committed:
treefmt: fix whitespace error in `withConfig` docs

MattSturgeon commented:
@jfly thanks for the review! I just pushed a fixup commit that breaks the module into smaller modules. LMK if you think it's better with or without that and I'll drop or squash as appropriate :slightly_smiling_face:

jfly reviewed (approved):
LGTM!

MattSturgeon commented:
> I'd like to flesh the manual section out a bit more, as currently it is effectively just a blank page to put the generated docs. Ideally, the various `treefmt.*` functions would have docs and examples, although I'm hesitant to increase the scope of this PR by writing loads of documentation.
> 
> I don't see a good way to include options in the nix doc-comments or include doc-comments in the manual. Maybe the doc-comments should just link to the manual?

I've come up with something locally that includes the function's reference docs in the manual via `nixdoc`.

To support `nixdoc`, 17fefd7f14d686ceda54c215cc7590eec6a533e6 needs to be re-worked to move all three functions into one file.

That's because `nixdoc` generates one manual section per input file. Additionally, it relies on functions being within a plain attrset (or any -> attrs) file. It can't document things defined within more complex expressions.

<details><summary>Screenshot</summary>
<p>

![image](https://github.com/user-attachments/assets/aadc4d29-b436-486f-99bb-21a0dba4c3dc)

</p>
</details> 


Local changes compared to this PR's current state: https://github.com/MattSturgeon/nixpkgs/compare/ae0e31481bea242c234de3e88f22e54d00c3d60a...95e93c5d06905c9ce5f3bc40b57fa59d09efa510

I need to think about whether to include that in this PR... The alternative is dropping the docs-related stuff from this PR and starting a follow-up PR that adds function docs and option docs in one go.

fricklerhandwerk commented:
Yeah I recommend merging what's good and following up with extra stuff later.

MattSturgeon committed:
treefmt: move `withConfig` and `buildConfig` into common file

This adds support for `:doc` in the nix repl and allows us to generate
docs for the functions using `nixdoc` if we choose.

MattSturgeon commented:
I've dropped the docs related changes from this PR, and will handle that in a follow up PR instead.

This PR is now a subset of what @jfly approved.

One tweak I've made in prep for a docs PR, is instead of moving doc-comments into the `with-config.nix` and `build-config.nix` files, I've combined them into a single `lib.nix` file. This will enable later using `nixdoc` to generate.

khaneliman reviewed (approved):
Makes sense and looks like a nice improvement to utilize the module system for this UX
